pkg/kubelet/cm/memorymanager/policy_static.go

```diff
@@ -97,29 +98,328 @@ func (p *staticPolicy) Start(ctx context.Context, s state.State) error {
+// shared pool. Such a configuration is invalid because it would lead to
+// containers in the shared pool having no memory allocated, causing them to
+// fail.
+func (p *staticPolicy) validatePodScopeResources(ctx context.Context, pod *v1.Pod) error {
+	// This will never happen, added to be explicit about this requirement
+	if p.affinity.Name() != topologymanager.PodTopologyScope {
+		return nil
+	}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResourceManagers) || !resourcehelper.IsPodLevelResourcesSet(pod) {
+		return nil
+	}
+
+	podTotalMemory, err := getPodRequestedResources(ctx, pod)
+	if err != nil {
+		return err
+	}
+
+	sumOfLongRunningExclusiveMemory := make(map[v1.ResourceName]uint64)
+	// Sum memory for all application containers.
+	for _, c := range pod.Spec.Containers {
+		containerReqs, err := getRequestedResources(ctx, pod, &c)
+		if err != nil {
+			return err
+		}
+		for resourceName, val := range containerReqs {
+			sumOfLongRunningExclusiveMemory[resourceName] += val
+		}
+	}
+	// Add memory for restartable init containers (sidecars), as they run for the full pod lifetime.
+	for _, c := range pod.Spec.InitContainers {
+		if podutil.IsRestartableInitContainer(&c) {
+			containerReqs, err := getRequestedResources(ctx, pod, &c)
+			if err != nil {
+				return err
+			}
+			for resourceName, val := range containerReqs {
+				sumOfLongRunningExclusiveMemory[resourceName] += val
+			}
+		}
+	}
+
+	// Check for empty shared pool for concurrently running containers (app and restartable init containers).
+	hasSharedAppContainers := false
+	for _, container := range pod.Spec.Containers {
+		req, err := getRequestedResources(ctx, pod, &container)
+		if err != nil {
+			return err
+		}
+		if len(req) == 0 {
+			hasSharedAppContainers = true
+			break
+		}
+	}
+	if !hasSharedAppContainers {
+		for _, container := range pod.Spec.InitContainers {
+			if !podutil.IsRestartableInitContainer(&container) {
+				continue
+			}
+			req, err := getRequestedResources(ctx, pod, &container)
+			if err != nil {
+				return err
+			}
+			if len(req) == 0 {
+				hasSharedAppContainers = true
+				break
+			}
+		}
+	}
+
+	if hasSharedAppContainers {
+		for resourceName, totalPodResource := range podTotalMemory {
+			if sumOfLongRunningExclusiveMemory[resourceName] >= totalPodResource {
+				return fmt.Errorf("pod rejected, sum of exclusive container %s requests equals pod budget, leaving no memory for shared containers", resourceName)
+			}
+		}
+	}
+
+	// Check for empty shared pool for standard init containers.
+	sumOfSidecarExclusiveMemory := make(map[v1.ResourceName]uint64)
+	for _, container := range pod.Spec.InitContainers {
+		if podutil.IsRestartableInitContainer(&container) {
+			req, err := getRequestedResources(ctx, pod, &container)
+			if err != nil {
+				return err
+			}
+			for resourceName, val := range req {
+				sumOfSidecarExclusiveMemory[resourceName] += val
+			}
+		}
+	}
```

Comment:
I realized this just now, but the following applies also for cpumanager: why do we need this check at all, again?
init containers run one at time serial, and their resources request cannot exced the pod bubble, by definition (and which is validated elsewhere). Then how can a init container ever get into the shared pool exhaustion?

Reply:
It is for the same reason as the standard containers, we could have an scenario where the amount of sidecars is equal to the pod-level resources, thus creating and empty pod shared pool for all non-guaranteed init containers.